### ISSUE TITLE
Cross Validation command

### ISSUE THREAD

**Issue by mhammadpk** [wnd-charm/wnd-charm#106]
Hi all,

I want to perform 5-Fold cross validation. I have used the following command to perform it:

wndchrm test -r0.2 -n5 featurefile.txt report.htm

Will the above command works such as:

1. Make 5 different partitions of whole dataset. 
2. Each time 1 partition is used for testing and 4 remaining partitions for training. 
3. Next time it will take another partition for testing and remaining others for training, and so on.
4. Report will contain 5 different classification results,  because each time combination of training and testing sets are different.

Please let me know about it. I will be thankful to you.

**Comment by colettace** [comment]
Hi,

Sorry, we haven't implemented n-fold cross validation in either the command line WND-CHARM implementation or the Python API.

You can write a script to do it using the Python API. Try something like this:

```
from wndcharm.FeatureSpace import FeatureSpace
from wndcharm.FeatureWeights import FisherFeatureWeights
from wndcharm.FeatureSpacePrediction import FeatureSpaceClassification
from wndcharm.FeatureSpacePredictionExperiment import FeatureSpaceClassificationExperiment

if __name__ == '__main__':
    import sys
    fname = sys.argv[1]
    print "Input filename:", fname
    fs = FeatureSpace.NewFromFitFile( fname, long=True )
    number_of_folds = 5
    feature_usage_fraction = 0.2 # Use top 20% ranked features for each iter

    exp = FeatureSpaceClassificationExperiment( training_set=fs, test_set=fs )
    for fold_index in xrange( number_of_folds ):
        test_samples = []
        # iterate over all the classes and obtain sample group ids for the nth fold of samples
        for inclass_sg_id_list in fs.sample_group_ids:
            num_samples_in_this_class = len( inclass_sg_id_list )
            num_samples_in_fold = num_samples_in_this_class / number_of_folds
            start_index = fold_index * num_samples_in_fold
            stop_index = ( fold_index + 1 ) * num_samples_in_fold
            test_samples.extend( inclass_sg_id_list[ start_index : stop_index ] )

        train_set = fs.SampleReduce( leave_out_sample_group_ids=test_samples )
        test_set = fs.SampleReduce( leave_in_sample_group_ids=test_samples )
        train_set.Normalize( inplace=True )
        weights = FisherFeatureWeights.NewFromFeatureSpace( train_set ).Threshold( feature_usage_fraction )
        train_set.FeatureReduce( weights, inplace=True )
        test_set.FeatureReduce( weights, inplace=True )
        test_set.Normalize( train_set )

        split_name = "fold" + str( fold_index ).zfill(2)
        split = FeatureSpaceClassification.NewWND5( train_set, test_set, weights, name=split_name)
        exp.individual_results.append( split )

    exp.Print()
```

Put that in a file called nfold.py and run it with a command like:
<code>python nfold.py featurefile.txt</code>

You may have to rename featurefile.txt to featurefile.fit or else the Python API may refuse to read it.

Let me know how it goes!
-Chris

**Comment by igg** [comment]
> On Mar 22, 2017, at 02:06, mhammadpk <notifications@github.com> wrote:
> 
> Hi all,
> 
> I want to perform 5-Fold cross validation. I have used the following command to perform it:
> 
> wndchrm test -r0.2 -n5 featurefile.txt report.htm
> 
Strictly speaking, (most) people who say "k-fold cross validation" mean something slightly different than what wndchrm does. The command above will result in 5 independent cross-validations: 5 training sets, 5 test sets. Each time, the training set is guaranteed to not overlap with the test set, so each one is a cross-validation. But, there are no guarantees about overlap between the 5 separate trials - they are fully independent random samples.

> Will the above command works such as:
> 
> Make 5 different partitions of whole dataset.
Yes.
> Each time 1 partition is used for testing and 4 remaining partitions for training.
Each of the 5 splits will result in one training set and one test set.
> Next time it will take another partition for testing and remaining others for training, and so on.
Second time will be a completely independent random split into training and testing samples. Independent means that samples are selected without regard to wether they were selected in other splits. The testing/training fractions refer to the whole dataset (independently of -n).
> Report will contain 5 different classification results, because each time combination of training and testing sets are different.
That is correct. Also, a summary of the 5 (mean, variance, etc).
> Please let me know about it. I will be thankful to you.
> 
The r of 0.2 means that the training fraction will be quite small (20%) in each split. Typically, datasets are small enough that training samples are a premium, so a larger r is used (75% training by default). To compensate for a small test sample, many independent rounds are done (-n) to get stable statistics that are descriptive of the dataset as a whole.

Keep in mind that by default the training is done with an equal number of samples per class.  So in each split, the r fraction will be applied to the smallest class you have and that number of samples will be used for training all the classes. The same logic applies to the test set.  This is to avoid biasing the classification result by your population structure, which is very important.  This balanced training/balanced testing requirement makes experiments like "leave one out" as well as a straight-forward k-fold cross-validation difficult to orchestrate because typically the dataset doesn't have the same number of samples in each class, so the data can't be split equally between classes and between subsets at the same time, so you end up either not using all the data, doing exhaustive permutations, or reverting back to random sampling. We went with random sampling because it was the most universal approach. It can sometimes be wasteful because getting stable statistics with small sets means running many simulations. For the WND5 classifier, training and testing is fast enough not to matter though.

-Ilya

**Comment by mhammadpk** [comment]
Hi all,

@ Chris

Thanks for sharing a script of cross validation. I have run the script, it works without errors. But, it prints all the results on Terminal. Is there a way to export the results as report in **HTML** format, just we do in command line utility of wndchrm? The report should be similar as wndchrm C++ program has generated, showing the results of each split and also the average results. Please let me know about it.

@Illya 
Thanks for your great explanation of split implementation in wndchrm.

**Comment by colettace** [comment]
Hi,

Sorry, HTML representation of results is not implemented for scripts. You'll have to make your own.

I recommend import the classification results into the Python package Pandas. Once the data is in, you can save it to html using the to_html() method, something like this:

```
import pandas as pd
exp.PerSampleStatistics( output_filepath='temp.tsv', print_indiv=False )
wc_results = pd.read_csv( temp.tsv, sep='\t' )
wc_results.to_html()
```

See [this stack overflow link](http://stackoverflow.com/questions/18096748/pandas-dataframes-to-html-highlighting-table-rows) for more info about writing HTML using Pandas.

Cheers,
Chris
